File: articles/probability-variation-biological-inference/python/permutation_test.py
"""
Permutation test workflow.

Run:
    python python/permutation_test.py
"""

from pathlib import Path

import pandas as pd

from probability_core import permutation_test_mean_difference


ARTICLE_DIR = Path(__file__).resolve().parents[1]
MEASURE_PATH = ARTICLE_DIR / "data" / "biological_measurements.csv"


def main() -> None:
    data = pd.read_csv(MEASURE_PATH)

    control = data.loc[data["group"] == "control", "value"].to_numpy()
    treated = data.loc[data["group"] == "treated", "value"].to_numpy()

    result = permutation_test_mean_difference(control, treated, n_permutations=10000, seed=42)

    print(pd.DataFrame([result]).round(6).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/probability-variation-biological-inference/r/permutation_test.R
# Permutation test workflow in R.

measure_path <- file.path("data", "biological_measurements.csv")
if (!file.exists(measure_path)) {
  measure_path <- file.path("..", "data", "biological_measurements.csv")
}

set.seed(42)

data <- read.csv(measure_path)

control <- data$value[data$group == "control"]
treated <- data$value[data$group == "treated"]

observed_difference <- mean(treated) - mean(control)
combined <- c(control, treated)
n_control <- length(control)

n_permutations <- 10000

permuted <- replicate(n_permutations, {
  shuffled <- sample(combined, replace = FALSE)
  mean(shuffled[(n_control + 1):length(shuffled)]) - mean(shuffled[1:n_control])
})

p_value <- mean(abs(permuted) >= abs(observed_difference))

summary_df <- data.frame(
  observed_difference = observed_difference,
  permutation_p_value = p_value,
  null_mean = mean(permuted),
  null_sd = sd(permuted)
)

print(round(summary_df, 6))
